linker/symbols.go

```diff
@@ -76,61 +92,100 @@ func (s *Symbols) Import(fd protoreflect.FileDescriptor, handler *reporter.Handl
 		}
 	}
 
-	if res, ok := fd.(*result); ok {
-		return s.importResult(res, true, handler)
+	if res, ok := fd.(*result); ok && res.hasSource() {
+		return s.importResultWithExtensions(pkg, res, handler)
 	}
 
-	if err := s.importPackages(sourcePositionForPackage(fd), fd.Package(), handler); err != nil {
+	return s.importFileWithExtensions(pkg, fd, handler)
+}
+
+func (s *Symbols) importFileWithExtensions(pkg *packageSymbols, fd protoreflect.FileDescriptor, handler *reporter.Handler) error {
+	imported, err := pkg.importFile(fd, handler)
+	if err != nil {
 		return err
 	}
+	if !imported {
+		// nothing else to do
+		return nil
+	}
+
+	return walk.Descriptors(fd, func(d protoreflect.Descriptor) error {
+		fld, ok := d.(protoreflect.FieldDescriptor)
+		if !ok || !fld.IsExtension() {
+			return nil
+		}
+		pos := sourcePositionForNumber(fld)
+		extendee := fld.ContainingMessage()
+		if err := s.AddExtension(extendee.ParentFile().Package(), extendee.FullName(), fld.Number(), pos, handler); err != nil {
+			return err
+		}
+		return nil
+	})
+}
 
+func (s *packageSymbols) importFile(fd protoreflect.FileDescriptor, handler *reporter.Handler) (bool, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	if _, ok := s.files[fd]; ok {
 		// have to double-check if it's already imported, in case
 		// it was added after above read-locked check
-		return nil
+		return false, nil
 	}
 
 	// first pass: check for conflicts
 	if err := s.checkFileLocked(fd, handler); err != nil {
-		return err
+		return false, err
 	}
 	if err := handler.Error(); err != nil {
-		return err
+		return false, err
 	}
 
 	// second pass: commit all symbols
 	s.commitFileLocked(fd)
 
-	return nil
+	return true, nil
 }
 
-func (s *Symbols) importPackages(pos ast.SourcePos, pkg protoreflect.FullName, handler *reporter.Handler) error {
+func (s *Symbols) importPackages(pos ast.SourcePos, pkg protoreflect.FullName, handler *reporter.Handler) (*packageSymbols, error) {
```

Comment:
nit: It would be nice to call the first parameter `pkgPos` to match other places in the code.

Reply:
done